Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm

```diff
@@ -363,8 +392,47 @@
         ASSERT_NOT_REACHED();
         return;
     }
+    auto replacedRange = resolvedRange;
+
+    auto subCharacterRange = [](WebCore::CharacterRange superRange, WebCore::CharacterRange previousRange) {
+        auto location = previousRange.location + previousRange.length;
+        return WebCore::CharacterRange(location, superRange.length - previousRange.length);
+    };
+
+    for (auto session : m_textIndicatorCharacterRangesForSessions) {
+        if (session.first == uuid) {
+            auto replaceCharacterRange = subCharacterRange(range, session.second.last().second);
+            replacedRange = WebCore::resolveCharacterRange(*sessionRange, replaceCharacterRange);
+        }
+    }
+
+    auto replaceTextIndicatorUUID = WTF::UUID::createVersion4();
+    m_webPage->createTextIndicatorForRange(replacedRange, [uuid = replaceTextIndicatorUUID, weakWebPage = WeakPtr { m_webPage }](std::optional<WebCore::TextIndicatorData>&& textIndicatorData) {
+        auto protectedWebPage = weakWebPage.get();
+        if (textIndicatorData)
+            protectedWebPage->addTextIndicatorStyleForID(uuid, WebKit::TextIndicatorStyle::Source, *textIndicatorData);
+    });
 
     replaceContentsOfRangeInSession(uuid, resolvedRange, *fragment);
+
+    auto sessionRangeAfterReplace = contextRangeForSessionWithUUID(uuid);
+    if (!sessionRange) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+    auto characterRangeAfterReplace = WebCore::CharacterRange(range.location, range.length + characterCountDelta);
+
+    bool attachedToExistingSession = false;
+    for (auto& session : m_textIndicatorCharacterRangesForSessions) {
+        if (session.first == uuid) {
+            session.second.append({ replaceTextIndicatorUUID, subCharacterRange(range, session.second.last().second) });
```

Comment:
what ensures that `session.second` is not empty?

Reply:
Why does it matter if the vector is empty, I'm just appending to it...

Reply:
if `session.second` is empty, then calling `last()` on it will crash afaik

Reply:
Oh, it's never empty because of how it's added to.